### PR TITLE
Custom PK support

### LINKED ISSUE #80: We assume the DB uses a column named `id`
We should be able to configure that per resource. For example, on one of our projects it's called `Id`

### PR DESCRIPTION
**Pull request by spersico**
closes #80 .

----------------
#### This PR should be merged AFTER my previous ones, as this touches a fair amount of files.

-------------

So far, this PR adds support for CRUD in a Resource which has a new schema attribute: _primaryKeyName_.

For example, a resource could be defined like this:
image

A GET to `/votes/1` brings this:
image

And a POST to `/votes` request and response would be:
image

### CODE REVIEW

File: src/application.ts [hunk 1]
```diff
@@ -193,7 +193,7 @@ export default class Application {
     Object.keys(data.relationships).forEach(relationshipName => {
       const relationships = this.serializeRelationship((data.relationships[
         relationshipName
-      ] as unknown) as Resource | Resource[]);
+      ] as unknown) as Resource | Resource[], schemaRelationships[relationshipName].type().schema.primaryKeyName);
```
Comment (joelalejandro) [inline src/application.ts]: @spersico Can we create some variables to make these lines a bit more... readable? 🙏

File: src/processors/knex-processor.ts [hunk 2]
```diff
@@ -147,18 +151,18 @@ export default class KnexProcessor<
...
+      .where({ [primaryKeyName]: id })
       .select(getColumns(this.resourceClass))
       .first();
   }
 
   async add(op: Operation): Promise<HasId> {
     const { type } = op.ref;
-
-    const ids = await this.getQuery().insert(op.data.attributes, "id");
+    const primaryKeyName = this.resourceClass.schema.primaryKeyName || "id";
```
Comment (joelalejandro) [inline src/processors/knex-processor.ts]: @spersico Seems like we could make the `schema` assume `id` has default so this defaulting can be done only once?

File: src/processors/knex-processor.ts [hunk 3]
```diff
@@ -245,24 +249,32 @@ export default class KnexProcessor<
...
     const query = relationProcessor.getQuery();
     const foreignTableName = relationProcessor.tableName;
     const sqlOperator = Array.isArray(result) ? "in" : "=";
+
+    const primaryKey = this.resourceClass.schema.primaryKeyName || "id";
+
     const queryIn: string | string[] = Array.isArray(result)
-      ? result.map((a: Resource) => a.id)
-      : result.id;
+      ? result.map((a: Resource) => a[primaryKey])
```
Comment (joelalejandro) [inline src/processors/knex-processor.ts]: @spersico can we name this variable `resource`? `a` means nothing in this context...

File: src/application.ts [hunk 4]
```diff
@@ -53,6 +54,7 @@ export default class Application {
     processor: OperationProcessor<Resource>
   ): Promise<OperationResponse> {
     const result = await processor.execute(await this.deserializeResource(op));
+    console.log('result', result)
```
Comment (joelalejandro) [inline src/application.ts]: @spersico This should probably go away

File: src/application.ts [hunk 5]
```diff
@@ -74,10 +76,8 @@ export default class Application {
           op.data.relationships.hasOwnProperty(relName)
       )
       .reduce((relationAttributes, relName) => {
-        let key =
-          schemaRelationships[relName].foreignKeyName ||
-          `${schemaRelationships[relName].type().type}Id`;
-        let value = (<ResourceRelationshipData>(
+        const key = schemaRelationships[relName].foreignKeyName || `${relName}Id`;
```
Comment (joelalejandro) [inline src/application.ts]: @spersico In another PR we should set a function to build `${relName}Id`, so can have in a single place the builder for that string.

File: src/application.ts [hunk 6]
```diff
@@ -189,13 +182,18 @@ export default class Application {
...
-
-    Object.keys(data.relationships).forEach(relationshipName => {
-      const relationships = this.serializeRelationship((data.relationships[
-        relationshipName
-      ] as unknown) as Resource | Resource[], schemaRelationships[relationshipName].type().schema.primaryKeyName);
-      data.relationships[relationshipName] = {
-        data: relationships
+    Object.keys(data.relationships).forEach(relName => {
+      const fkName = schemaRelationships[relName].belongsTo
+        ? "id"
```
Comment (joelalejandro) [inline src/application.ts]: @spersico Shouldn't this use `DEFAULT_PRIMARY_KEY`?

File: src/application.ts [hunk 7]
```diff
@@ -225,7 +223,7 @@ export default class Application {
     if (!data) {
       return null;
     }
-
+    console.log('data', data)
```
Comment (joelalejandro) [inline src/application.ts]: @spersico This should probably go away

File: src/application.ts (hunk 6 above)
Comment (spersico) [reply src/application.ts]: Noup, in line 168 we build the resources object using "id" from the proper FK. 
This forEach ensures only that all elements in the relationships object have ONLY an **id** and **type** attributes, inside a data object.

File: src/application.ts (hunk 5 above)
Comment (spersico) [reply src/application.ts]: created an issue with this

**Pull request merged**